### PR TITLE
Suggestions for Non-integer binning

### PR DESCRIPTION
**Pull request by thomasaarholt**
Please take a look. As far as I can see, it works. I do have some issues with the function in general though - the function seems to leave zero-lines here and there. May have to do with math.floor/ceil and having chosen the wrong of the two here or there.

screenshot 2016-10-12 22 20 03

### CODE REVIEW

**Comment by thomasaarholt** [comment]
I may have mixed up which of the scale factors bin which dimension. Looks like [1,1,4] bins the second axis by 4.

**Comment by k8macarthur** [comment]
So you may end up with a black line around the edge for the remainder of the counts that have less than one pixel. This could be later cropped out but should be kept so that you can check that counts are preserved. It might take me a while to work out what you've changed but I'll try and work my way through and message when I have comments.

**Comment by k8macarthur** [comment]
Is the function still callable when you have it signals rather than EDS?

File: hyperspy/_signals/eds.py [hunk 1]
```diff
@@ -233,109 +233,6 @@ def rebin(self, new_shape):
                     *= factor
         return s
 
-    def linear_bin(self, scale):
```
Comment (k8macarthur) [inline hyperspy/_signals/eds.py:236]: Does the function still get called properly from signals rather than EDS?

**Comment by k8macarthur** [comment]
Wow there seem to be rebin functions all over the place! Does moving the function to base signal mean it's accessible from everywhere else?

**Comment by k8macarthur** [comment]
I have a location rejig so that now there is only one function in array_tools which can then be called elsewhere for different class types. Are you able to update so I can see your other changes?

**Comment by thomasaarholt** [comment]
That's correct. Think of BaseSignal as a "parent" class. All Signals (Signal1D, Signal2D, ComplexSignal, etc.) inherit BaseSignal's methods. The EDS signal is child of Signal1D, so contains BaseSignal's methods, Signal1Ds methods as well as the specific EDS ones. And so on for EDS_TEM and EDS_SEM.

File: hyperspy/_signals/eds.py (hunk 1 above)
Comment (thomasaarholt) [reply hyperspy/_signals/eds.py:236]: If you mean, is an EDS signal able to call the method - after moving it to BaseSignal from EDS - then yes :)

**Comment by thomasaarholt** [comment]
Make sure you look at the commits I've done to get a "step-by-step" view of what I changed. I tried to make it as clear as possible.

**Comment by thomasaarholt** [comment]
To be clear in what I mean by checking out the commits - click the highlighted link above (in blue on the attached image), and then step through using the "next" button in the top right to see each commit. I highly advise making a commit each time you add something so that it's clear what the change is for others. (rather than piling up a lot of changes and commiting them at the same time)
screenshot 2016-10-13 16 24 06

**Comment by k8macarthur** [comment]
Thanks Tom I've been trying to look through at all the step changes  like you suggested. I probably should have gone through your changes first before doing my rearrange. I'm seeing if I can remember how to cherry pick now.

**Comment by thomasaarholt** [comment]
Good luck :p

**Comment by thomasaarholt** [comment]
Actually, I would still argue that this should be a method under BaseSignal - I can't think of a single type of signal that wouldn't benefit from the method?

**Comment by thomasaarholt** [comment]
Bug for Signal2Ds: The bug I mentioned previously is only for Signal2Ds - it's because the s.data.shape lists the signal dimensions in reverse order to the navigation dimensions.
So a signal hyperspy shape `(N | X Y)` has `s.data.shape` `(N, Y, X)`. Since the iterator doesn't know that, it bins incorrectly. I'll have a think.

**Comment by k8macarthur** [comment]
Ok I think I've accepted everything other than your switch around the if statement. Let me know what you think. I hope you got authorship but feel free to add subsequent small changes. I think the reason this was difficult to work out because you did a pull request for to many changes and I think the if statement change might have been the route of the problem with your axis binning in the wrong order. Feel free to check out some more and let me know.

**Comment by k8macarthur** [comment]
I think separately calling linear_bin from each function is the way to go rather than in signals.py because of the separate requirements for changing the dwell_time etc.

**Comment by thomasaarholt** [comment]
Cool - I'll test the new pull request and see if the signal2D-bug is still present. I have an idea on how to fix it, if it is present. I'm closing this PR now. Will open a new one if I have more suggestions.

I had not thought about the dwell time problem (that you just mentioned). You're right.

**Pull request closed without merging**